# Pandas

Материалы:
* Макрушин С.В. "Лекция 2: Библиотека Pandas"
* https://pandas.pydata.org/docs/user_guide/index.html#
* https://pandas.pydata.org/docs/reference/index.html
* Уэс Маккини. Python и анализ данных

## Задачи для совместного разбора

In [1]:
import numpy as np
import pandas as pd

1. Загрузите данные из файла `sp500hst.txt` и обозначьте столбцы в соответствии с содержимым: `"date", "ticker", "open", "high", "low", "close", "volume"`.

In [2]:
data = pd.read_csv('sp500hst.txt', names = ["date", "ticker", "open", "high", "low", "close", "volume"])

In [3]:
data.head()

date ticker   open     high     low  close  volume
0  20090821      A  25.60  25.6100  25.220  25.55   34758
1  20090824      A  25.64  25.7400  25.330  25.50   22247
2  20090825      A  25.50  25.7000  25.225  25.34   30891
3  20090826      A  25.32  25.6425  25.145  25.48   33334
4  20090827      A  25.50  25.5700  25.230  25.54   70176

2. Рассчитайте среднее значение показателей для каждого из столбцов c номерами 3-6.

In [4]:
for i in range(3,7):
    print(f'Средний показатель {data.columns[i]} = {data.iloc[:, i].mean()}')

Средний показатель high = 43.102243387667855
Средний показатель low = 42.05446366521448
Средний показатель close = 42.60186484817335
Средний показатель volume = 81395.06813843067


3. Добавьте столбец, содержащий только число месяца, к которому относится дата.

In [5]:
data['month'] = data['date'].apply(lambda x: int(str(x)[5:6]) if str(x)[4] == '0' else int(str(x)[4:6]))

In [6]:
data.sample(10)

date ticker   open     high    low  close  volume  month
25238   20100805    COH  39.00  39.0000  38.50  38.73   42272      8
48209   20090916    GPC  36.08  36.7600  35.94  36.72    6730      9
103077  20100714    STT  38.11  38.2900  37.39  37.58   56346      7
54773   20100119    HSY  37.26  38.0100  37.26  37.35   35437      1
102143  20100819    SRE  50.93  51.1000  50.16  50.60   11959      8
21487   20100521   CHRW  56.53  58.2400  56.26  58.18   28852      5
86003   20091203   PDCO  26.40  26.5300  25.95  25.98    9606     12
102615  20100723    STJ  35.43  36.9700  35.43  36.81   57473      7
107782  20090929    TJX  37.44  37.6705  36.83  37.00   28775      9
12261   20100422    BBT  33.50  34.3600  33.01  34.34  116642      4

4. Рассчитайте суммарный объем торгов для для одинаковых значений тикеров.

In [7]:
data[['ticker', 'volume']].groupby(['ticker']).sum()

volume
ticker          
A        8609336
AA      81898998
AAPL    52261170
ABC      9006756
ABT     18975870
...          ...
XTO     21297931
YHOO    56837171
YUM     10971538
ZION    15551119
ZMH      4938916

[524 rows x 1 columns]

5. Загрузите данные из файла sp500hst.txt и обозначьте столбцы в соответствии с содержимым: "date", "ticker", "open", "high", "low", "close", "volume". Добавьте столбец с расшифровкой названия тикера, используя данные из файла `sp_data2.csv` . В случае нехватки данных об именах тикеров корректно обработать их.

In [59]:
arr1 = pd.read_csv('sp500hst.txt', names = ["date", "ticker", "open", "high", "low", "close", "volume"])
arr1.sample(5)

date ticker   open   high    low  close  volume
19420   20100120    CCE  20.95  21.21  20.85  21.16   37499
77588   20091027    NEM  43.33  43.78  42.75  43.28   74661
109440  20100713    TSN  17.71  17.95  17.64  17.74   43527
40131   20100204   EXPD  33.76  33.98  33.42  33.42   26233
37188   20100510    EMN  64.33  65.56  64.06  65.44   16363

In [60]:
arr2 = pd.read_csv('sp_data2.csv', delimiter=';', names=['ticker', 'company', 'procent'])
arr2.head(5), arr1.shape, arr2.shape

(  ticker     company procent
 0   AAPL       Apple    3.6%
 1   AMZN  Amazon.com    3.2%
 2  GOOGL    Alphabet    3.1%
 3   GOOG    Alphabet    3.1%
 4   MSFT   Microsoft    3.0%,
 (122574, 7),
 (505, 3))

In [61]:
arr2['procent']=arr2['procent'].apply(lambda x: float(x[:-1]))
arr2

ticker          company  procent
0     AAPL            Apple      3.6
1     AMZN       Amazon.com      3.2
2    GOOGL         Alphabet      3.1
3     GOOG         Alphabet      3.1
4     MSFT        Microsoft      3.0
..     ...              ...      ...
500    SCG            SCANA      0.0
501    AIZ         Assurant      0.0
502    AYI    Acuity Brands      0.0
503    HRB        H&R Block      0.0
504    RRC  Range Resources      0.0

[505 rows x 3 columns]

In [62]:
arr3 = arr2.groupby(['ticker', 'company'], as_index=False).mean()
arr3

ticker                  company  procent
0        A     Agilent Technologies      0.1
1      AAL  American Airlines Group      0.1
2      AAP       Advance Auto Parts      0.0
3     AAPL                    Apple      3.6
4     ABBV                   AbbVie      0.6
..     ...                      ...      ...
500    XYL                    Xylem      0.0
501    YUM               Yum Brands      0.1
502    ZBH   Zimmer Biomet Holdings      0.1
503   ZION            Zions Bancorp      0.0
504    ZTS                   Zoetis      0.2

[505 rows x 3 columns]

In [63]:
dictionary = dict(zip(arr3.ticker.values, arr3.company.values))
arr1['company'] = arr1['ticker'].map(dictionary)

In [98]:
arr1.tail(10)

date ticker   open   high    low  close  volume company
122564  20100806    ZMH  54.08  54.30  53.26  53.98   15890     NaN
122565  20100809    ZMH  54.39  54.49  53.72  53.99   12170     NaN
122566  20100810    ZMH  53.61  54.40  53.29  53.90   21266     NaN
122567  20100811    ZMH  53.21  53.21  51.89  52.01   33017     NaN
122568  20100812    ZMH  51.45  52.32  51.31  52.00   28473     NaN
122569  20100813    ZMH  51.72  51.90  51.38  51.44   14561     NaN
122570  20100816    ZMH  51.13  51.47  50.60  51.00   13489     NaN
122571  20100817    ZMH  51.14  51.60  50.89  51.21   20498     NaN
122572  20100819    ZMH  51.63  51.63  50.17  50.22   18259     NaN
122573  20100820    ZMH  50.03  50.55  49.48  49.82   17792     NaN

In [91]:
# нераспознанные ticker в arr1
arr1.iloc[np.where(arr1['company'].isna())[0],:]['ticker'].unique()

array(['AA', 'ACE', 'ACS', 'AKS', 'ALTR', 'AN', 'ANF', 'AOC', 'APOL',
       'ARG', 'ATI', 'AVP', 'AYE', 'BBBY', 'BCR', 'BDK', 'BHI', 'BIG',
       'BJS', 'BMC', 'BMS', 'BNI', 'BRCM', 'BTU', 'CAM', 'CBE', 'CBG',
       'CCE', 'CEG', 'CEPH', 'CFN', 'CHK', 'CIEN', 'CLF', 'CNX', 'COH',
       'CPWR', 'CSC', 'CVG', 'CVH', 'DD', 'DELL', 'DF', 'DNB', 'DNR',
       'DO', 'DOW', 'DTV', 'DV', 'DYN', 'EK', 'EMC', 'EP', 'ERTS', 'ESV',
       'ETFCD', 'FDO', 'FHN', 'FII', 'FO', 'FPL', 'FRX', 'FSLR', 'FTR',
       'GAS', 'GCI', 'GENZ', 'GME', 'GNW', 'GR', 'HAR', 'HCBK', 'HCN',
       'HNZ', 'HOT', 'HSP', 'IGT', 'ITT', 'JAVA', 'JBL', 'JCP', 'JDSU',
       'JNS', 'KBH', 'KFT', 'KG', 'LIFE', 'LLTC', 'LM', 'LO', 'LSI',
       'LTD', 'LUK', 'LXK', 'MBI', 'MDP', 'MEE', 'MFE', 'MHP', 'MHS',
       'MI', 'MIL', 'MJN', 'MOLX', 'MON', 'MOT', 'MTW', 'MUR', 'MWV',
       'MWW', 'NBR', 'NOVL', 'NSM', 'NU', 'NVLS', 'NYT', 'NYX', 'ODP',
       'OI', 'PBG', 'PBI', 'PCL', 'PCLN', 'PCP', 'PCS', 'PDCO', 'PGN',
      

In [100]:
arr1.company.fillna('no information', inplace=True)

In [101]:
arr1.isna().sum()

date       0
ticker     0
open       0
high       0
low        0
close      0
volume     0
company    0
dtype: int64

In [103]:
arr1.tail()

date ticker   open   high    low  close  volume         company
122569  20100813    ZMH  51.72  51.90  51.38  51.44   14561  no information
122570  20100816    ZMH  51.13  51.47  50.60  51.00   13489  no information
122571  20100817    ZMH  51.14  51.60  50.89  51.21   20498  no information
122572  20100819    ZMH  51.63  51.63  50.17  50.22   18259  no information
122573  20100820    ZMH  50.03  50.55  49.48  49.82   17792  no information

## Лабораторная работа №2

### Базовые операции с `DataFrame`

1.1 В файлах `recipes_sample.csv` и `reviews_sample.csv` находится информация об рецептах блюд и отзывах на эти рецепты соответственно. Загрузите данные из файлов в виде `pd.DataFrame` с названиями `recipes` и `reviews`. Обратите внимание на корректное считывание столбца с индексами в таблице `reviews` (безымянный столбец).

In [112]:
recipes = pd.read_csv('recipes_sample.csv')
reviews = pd.read_csv('reviews_sample.csv', skiprows=1, names=['index', 'user_id', 'recipe_id',
                                                               'date', 'rating','review'])

In [118]:
reviews.head()

index     user_id  recipe_id        date  rating  \
0  370476       21752      57993  2003-05-01       5   
1  624300      431813     142201  2007-09-16       5   
2  187037      400708     252013  2008-01-10       4   
3  706134  2001852463     404716  2017-12-11       5   
4  312179       95810     129396  2008-03-14       5   

                                              review  
0  Last week whole sides of frozen salmon fillet ...  
1  So simple and so tasty!  I used a yellow capsi...  
2  Very nice breakfast HH, easy to make and yummy...  
3  These are a favorite for the holidays and so e...  
4  Excellent soup!  The tomato flavor is just gre...

In [114]:
recipes.head()

name     id  minutes  contributor_id  \
0     george s at the cove  black bean soup  44123       90           35193   
1        healthy for them  yogurt popsicles  67664       10           91970   
2              i can t believe it s spinach  38798       30            1533   
3                      italian  gut busters  35173       45           22724   
4  love is in the air  beef fondue   sauces  84797       25            4470   

    submitted  n_steps                                        description  \
0  2002-10-25      NaN  an original recipe created by chef scott meska...   
1  2003-07-26      NaN  my children and their friends ask for my homem...   
2  2002-08-29      NaN            these were so go, it surprised even me.   
3  2002-07-27      NaN  my sister-in-law made these for us at a family...   
4  2004-02-23      4.0  i think a fondue is a very romantic casual din...   

   n_ingredients  
0           18.0  
1            NaN  
2            8.0  
3            NaN  
4            NaN

In [119]:
reviews.shape, recipes.shape

((126696, 6), (30000, 8))

1.2 Для каждой из таблиц выведите основные параметры:
* количество точек данных (строк);
* количество столбцов;
* тип данных каждого столбца.

In [120]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126696 entries, 0 to 126695
Data columns (total 6 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   index      126696 non-null  int64 
 1   user_id    126696 non-null  int64 
 2   recipe_id  126696 non-null  int64 
 3   date       126696 non-null  object
 4   rating     126696 non-null  int64 
 5   review     126679 non-null  object
dtypes: int64(4), object(2)
memory usage: 5.8+ MB


In [121]:
recipes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   name            30000 non-null  object 
 1   id              30000 non-null  int64  
 2   minutes         30000 non-null  int64  
 3   contributor_id  30000 non-null  int64  
 4   submitted       30000 non-null  object 
 5   n_steps         18810 non-null  float64
 6   description     29377 non-null  object 
 7   n_ingredients   21120 non-null  float64
dtypes: float64(2), int64(3), object(3)
memory usage: 1.8+ MB


1.3 Исследуйте, в каких столбцах таблиц содержатся пропуски. Посчитайте долю строк, содержащих пропуски, в отношении к общему количеству строк.

In [129]:
# процент пропусков в отзывах
reviews.isna().sum()/reviews.shape[0] * 100

index        0.000000
user_id      0.000000
recipe_id    0.000000
date         0.000000
rating       0.000000
review       0.013418
dtype: float64

In [130]:
# процент пропусков в рецептах
recipes.isna().sum()/recipes.shape[0] * 100

name               0.000000
id                 0.000000
minutes            0.000000
contributor_id     0.000000
submitted          0.000000
n_steps           37.300000
description        2.076667
n_ingredients     29.600000
dtype: float64

1.4 Рассчитайте среднее значение для каждого из числовых столбцов (где это имеет смысл).

In [155]:
# средний рейтинг в отзывах
reviews[['rating']].mean()

rating    4.410802
dtype: float64

In [152]:
recipes[['minutes', 'n_steps', 'n_ingredients']].mean()

minutes          123.358133
n_steps            9.805582
n_ingredients      9.008286
dtype: float64

1.5 Создайте серию из 10 случайных названий рецептов.

In [166]:
pd.Series(recipes.name.sample(10)).values

array(['crock pot barbecue steak',
       'becky s southwest breakfast casserole',
       'slow cooker chicken sloppy joes', 'mushroom smothered chicken',
       'creamy mustard dressing', 'lisa s grilled steak marsala',
       'candied spiced mixed nuts  smucker s',
       'japanese inspired cucumber salad', 'carrot pep',
       'chicken bacon pie'], dtype=object)

1.6 Измените индекс в таблице `reviews`, пронумеровав строки, начиная с нуля.

In [172]:
reviews['index'] = np.arange(reviews.shape[0])

In [173]:
reviews

index     user_id  recipe_id        date  rating  \
0            0       21752      57993  2003-05-01       5   
1            1      431813     142201  2007-09-16       5   
2            2      400708     252013  2008-01-10       4   
3            3  2001852463     404716  2017-12-11       5   
4            4       95810     129396  2008-03-14       5   
...        ...         ...        ...         ...     ...   
126691  126691     1270706     335534  2009-05-17       4   
126692  126692     2282344       8701  2012-06-03       0   
126693  126693      689540     222001  2008-04-08       5   
126694  126694  2000242659     354979  2015-06-02       5   
126695  126695      463435     415599  2010-09-30       5   

                                                   review  
0       Last week whole sides of frozen salmon fillet ...  
1       So simple and so tasty!  I used a yellow capsi...  
2       Very nice breakfast HH, easy to make and yummy...  
3       These are a favorite for the holidays and so e...  
4       Excellent soup!  The tomato flavor is just gre...  
...                                                   ...  
126691  This recipe was great! I made it last night. I...  
126692  This recipe is outstanding. I followed the rec...  
126693  Well, we were not a crowd but it was a fabulou...  
126694  I have been a steak eater and dedicated BBQ gr...  
126695  Wonderful and simple to prepare seasoning blen...  

[126696 rows x 6 columns]

1.7 Выведите информацию о рецептах, время выполнения которых не больше 20 минут и кол-во ингредиентов в которых не больше 5.

In [187]:
recipes.loc[np.where((recipes.minutes <= 20) & (recipes.n_ingredients <=5))]

name      id  minutes  \
28                                   quick biscuit bread  302399       20   
60                         peas  fit for a king or queen  303944       20   
90                     hawaiian sunrise           mimosa  100837        5   
91            tasty dish s   banana pudding in 2 minutes  286484        2   
94                                    1 minute meatballs   11361       13   
...                                                  ...     ...      ...   
29873  zip and steam red potatoes with butter and garlic  304922       13   
29874                          ziplock vanilla ice cream   74250       10   
29905                      zucchini and corn with cheese  256177       15   
29980               zucchini with jalapeno monterey jack  320622       10   
29983                          zucchini with serrano ham  162411       15   

       contributor_id   submitted  n_steps  \
28             213909  2008-05-06     11.0   
60             213909  2008-05-16      NaN   
90              58104  2004-09-29      4.0   
91              47892  2008-02-13      NaN   
94               4470  2001-09-03      NaN   
...               ...         ...      ...   
29873          724218  2008-05-27      9.0   
29874           24386  2003-10-29      8.0   
29905          305531  2007-09-29      4.0   
29980          305531  2008-08-20      3.0   
29983          152500  2006-03-31      6.0   

                                             description  n_ingredients  
28     this is a wonderful quick bread to make as an ...            5.0  
60     this recipe is so simple and the flavors are s...            5.0  
90     pineapple mimosa was changed to hawaiian sunri...            3.0  
91     "mmmm, i love bananas!" a --tasty dish-- origi...            4.0  
94     this is a real short cut for cooks in a hurry....            2.0  
...                                                  ...            ...  
29873  i haven't tried this yet, but i am going to so...            5.0  
29874  a fun thing for kids to do. may want to use mi...            3.0  
29905  from betty crocker fresh spring recipes. i lik...            5.0  
29980                                  simple and yummy!            3.0  
29983  this dish is from tim malzer, a german chef wh...            5.0  

[2019 rows x 8 columns]

### Работа с датами в `pandas`

2.1 Преобразуйте столбец `submitted` из таблицы `recipes` в формат времени. Модифицируйте решение задачи 1.1 так, чтобы считать столбец сразу в нужном формате.

In [190]:
recipes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   name            30000 non-null  object 
 1   id              30000 non-null  int64  
 2   minutes         30000 non-null  int64  
 3   contributor_id  30000 non-null  int64  
 4   submitted       30000 non-null  object 
 5   n_steps         18810 non-null  float64
 6   description     29377 non-null  object 
 7   n_ingredients   21120 non-null  float64
dtypes: float64(2), int64(3), object(3)
memory usage: 1.8+ MB


In [210]:
recipes = pd.read_csv('recipes_sample.csv',converters={'submitted':pd.to_datetime})

In [211]:
recipes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   name            30000 non-null  object        
 1   id              30000 non-null  int64         
 2   minutes         30000 non-null  int64         
 3   contributor_id  30000 non-null  int64         
 4   submitted       30000 non-null  datetime64[ns]
 5   n_steps         18810 non-null  float64       
 6   description     29377 non-null  object        
 7   n_ingredients   21120 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(3), object(2)
memory usage: 1.8+ MB


2.2 Выведите информацию о рецептах, добавленных в датасет не позже 2010 года.

In [216]:
recipes.loc[recipes.submitted < '1/1/2011']

name      id  minutes  \
0             george s at the cove  black bean soup   44123       90   
1                healthy for them  yogurt popsicles   67664       10   
2                      i can t believe it s spinach   38798       30   
3                              italian  gut busters   35173       45   
4          love is in the air  beef fondue   sauces   84797       25   
...                                             ...     ...      ...   
29993                     zuni caf zucchini pickles  316950     2895   
29995  zurie s holey rustic olive and cheddar bread  267661       80   
29996          zwetschgenkuchen  bavarian plum cake  386977      240   
29997   zwiebelkuchen   southwest german onion cake  103312       75   
29999        cookies by design   cookies on a stick  298512       29   

       contributor_id  submitted  n_steps  \
0               35193 2002-10-25      NaN   
1               91970 2003-07-26      NaN   
2                1533 2002-08-29      NaN   
3               22724 2002-07-27      NaN   
4                4470 2004-02-23      4.0   
...               ...        ...      ...   
29993           62264 2008-07-31      NaN   
29995          200862 2007-11-25     16.0   
29996          177443 2009-08-24      NaN   
29997          161745 2004-11-03      NaN   
29999          506822 2008-04-15      9.0   

                                             description  n_ingredients  
0      an original recipe created by chef scott meska...           18.0  
1      my children and their friends ask for my homem...            NaN  
2                these were so go, it surprised even me.            8.0  
3      my sister-in-law made these for us at a family...            NaN  
4      i think a fondue is a very romantic casual din...            NaN  
...                                                  ...            ...  
29993  refrigerator pickles for some of the zucchini ...            8.0  
29995  this is based on a french recipe but i changed...           10.0  
29996  this is a traditional fresh plum cake, thought...           11.0  
29997  this is a traditional late summer early fall s...            NaN  
29999  i've heard of the 'cookies by design' company,...           10.0  

[27661 rows x 8 columns]

### Работа со строковыми данными в `pandas`

3.1  Добавьте в таблицу `recipes` столбец `description_length`, в котором хранится длина описания рецепта из столбца `description`.

3.2 Измените название каждого рецепта в таблице `recipes` таким образом, чтобы каждое слово в названии начиналось с прописной буквы.

3.3 Добавьте в таблицу `recipes` столбец `name_word_count`, в котором хранится количество слов из названии рецепта (считайте, что слова в названии разделяются только пробелами). Обратите внимание, что между словами может располагаться несколько пробелов подряд.

### Группировки таблиц `pd.DataFrame`

4.1 Посчитайте количество рецептов, представленных каждым из участников (`contributor_id`). Какой участник добавил максимальное кол-во рецептов?

4.2 Посчитайте средний рейтинг к каждому из рецептов. Для скольких рецептов отсутствуют отзывы? Обратите внимание, что отзыв с нулевым рейтингом или не заполненным текстовым описанием не считается отсутствующим.

4.3 Посчитайте количество рецептов с разбивкой по годам создания.

### Объединение таблиц `pd.DataFrame`

5.1 При помощи объединения таблиц, создайте `DataFrame`, состоящий из четырех столбцов: `id`, `name`, `user_id`, `rating`. Рецепты, на которые не оставлен ни один отзыв, должны отсутствовать в полученной таблице. Подтвердите правильность работы вашего кода, выбрав рецепт, не имеющий отзывов, и попытавшись найти строку, соответствующую этому рецепту, в полученном `DataFrame`.

5.2 При помощи объединения таблиц и группировок, создайте `DataFrame`, состоящий из трех столбцов: `recipe_id`, `name`, `review_count`, где столбец `review_count` содержит кол-во отзывов, оставленных на рецепт `recipe_id`. У рецептов, на которые не оставлен ни один отзыв, в столбце `review_count` должен быть указан 0. Подтвердите правильность работы вашего кода, выбрав рецепт, не имеющий отзывов, и найдя строку, соответствующую этому рецепту, в полученном `DataFrame`.

5.3. Выясните, рецепты, добавленные в каком году, имеют наименьший средний рейтинг?

### Сохранение таблиц `pd.DataFrame`

6.1 Отсортируйте таблицу в порядке убывания величины столбца `name_word_count` и сохраните результаты выполнения заданий 3.1-3.3 в csv файл. 

6.2 Воспользовавшись `pd.ExcelWriter`, cохраните результаты 5.1 и 5.2 в файл: на лист с названием `Рецепты с оценками` сохраните результаты выполнения 5.1; на лист с названием `Количество отзывов по рецептам` сохраните результаты выполнения 5.2.

#### [версия 2]
* Уточнены формулировки задач 1.1, 3.3, 4.2, 5.1, 5.2, 5.3